src/core/ipns/routing/pubsub.js

```diff
@@ -0,0 +1,206 @@
+    this._pubsub = pubsub
+
+    this._subscriptions = {}
+
+    // Bind _handleSubscriptionKey function, which is called by datastorePubsub.
+    this._handleSubscriptionKey = this._handleSubscriptionKey.bind(this)
+    this._dsPubsub = new DatastorePubsub(pubsub, localDatastore, peerId, ipns.validator, this._handleSubscriptionKey)
+  }
+
+  /**
+   * Put a value to the pubsub datastore indexed by the received key properly encoded.
+   * @param {Buffer} key identifier of the value.
+   * @param {Buffer} value value to be stored.
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  put (key, value, callback) {
+    if (!Buffer.isBuffer(key)) {
+      const errMsg = `key does not have a valid format`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    if (!Buffer.isBuffer(value)) {
+      const errMsg = `received value is not a buffer`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_VALUE_RECEIVED'))
+    }
+
+    this._dsPubsub.put(key, value, callback)
+  }
+
+  /**
+   * Get a value from the pubsub datastore indexed by the received key properly encoded.
+   * Moreover, the identifier topic is subscribed and the pubsub datastore records will be
+   * updated once new publishes occur.
+   * @param {Buffer} key identifier of the value to be obtained.
+   * @param {function(Error, Buffer)} callback
+   * @returns {void}
+   */
+  get (key, callback) {
+    if (!Buffer.isBuffer(key)) {
+      const errMsg = `key does not have a valid format`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    this._dsPubsub.get(key, (err, res) => {
+      // Add topic subscribed
+      const ns = key.slice(0, ipnsNSLength)
+
+      if (ns.toString() === ipnsNS) {
+        const stringifiedTopic = key.toString()
+        const id = toB58String(key.slice(ipnsNSLength))
+
+        this._subscriptions[stringifiedTopic] = id
+
+        log(`subscribed pubsub ${stringifiedTopic}: ${id}`)
+      }
+
+      if (err) {
+        return callback(err)
+      }
+
+      callback(null, res)
+    })
+  }
+
+  // Modify subscription key to have a proper encoding
+  // Without this, the utf-8 encoding gets the key broken
+  _handleSubscriptionKey (key, callback) {
+    const subscriber = this._subscriptions[key]
+
+    if (!subscriber) {
+      const errMsg = `key ${key} does not correspond to a subscription`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    let keys
+    try {
+      keys = ipns.getIdKeys(fromB58String(subscriber))
+    } catch (err) {
+      log.error(err)
+      return callback(err)
+    }
+
+    callback(null, keys.routingKey.toBuffer())
+  }
+
+  /**
+   * Get pubsub subscriptions related to ipns.
+   * @param {function(Error, Object)} callback
+   * @returns {void}
+   */
+  getSubscriptions (callback) {
```

Comment:
Why can't we just use `Object.keys(this._subscriptions)` here?

Reply:
We have different encodings between regular `pubsub` and `ipns pubsub` topics.

Test scenario:
```sh
> ipfs name resolve /ipns/QmWc7wfpSQf1rGLijVYAWxmVh7mSXzHsF8daqShaMCHeR9
> ipfs name pubsub subs
/ipns/QmWc7wfpSQf1rGLijVYAWxmVh7mSXzHsF8daqShaMCHeR9
> ipfs pubsub ls
/ipns/ z�?��)�%�o���c���M���!��-V͜
```

That is the reason why I added the `_subscriptions`, which maps the bytes from the real pubsub subscription with the friendly identifier for ipns pubsub users.